plumbing/transport/ssh/common.go

```diff
@@ -166,10 +167,33 @@ func dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
 	}
 	defer cancel()
 
-	conn, err := proxy.Dial(ctx, network, addr)
+	var conn net.Conn
+	var err error
+
+	if proxyOpts.URL != "" {
+		proxyUrl, err := proxyOpts.FullURL()
+		if err != nil {
+			return nil, err
+		}
+		dialer, err := proxy.FromURL(proxyUrl, proxy.Direct)
+		if err != nil {
+			return nil, err
+		}
+
+		// Try to use a ContextDialer, but fall back to a Dialer if that goes south.
+		ctxDialer, ok := dialer.(proxy.ContextDialer)
+		if ok {
+			conn, err = ctxDialer.DialContext(ctx, "tcp", addr)
+		} else {
+			conn, err = dialer.Dial("tcp", addr)
+		}
+	} else {
+		conn, err = proxy.Dial(ctx, network, addr)
+	}
```

Comment:
I view it as a tradeoff. With your suggested change, we would enforce the context but we would NOT use the custom proxy addr provided. IMO, its better to use the proxy and let go of the context.
Furthermore, this is just a fallback. looking at the actual object returned by `proxy.FromURL()`,  I don't think the `ctxDialer, ok := dialer.(proxy.ContextDialer)` type assertion will actually fail

Reply:
You are right, based on the execution path the `dialer` should always implement `proxy.ContextDialer`. So I think we are better off erroring when that isn't the case, so that ensures we always enforce the context timeout, but also can debug if things ever went awry:

```go
		ctxDialer, ok := dialer.(proxy.ContextDialer)
		if !ok {
			return nil, fmt.Errorf("...")
		}
		conn, err = ctxDialer.DialContext(ctx, "tcp", addr)
```